flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+                raise LookupError(f'The signac job {job.get_id()} not found'
+                                  f'in {project}')
+
+        for aggregate in aggregated_jobs:
+            try:
+                filter_aggregate = tuple(filter(_validate_and_filter_job, aggregate))
+            except TypeError:  # aggregate is not iterable
+                ValueError("Invalid aggregator_function provided by the user.")
+            # Store aggregate in this instance
+            self._aggregates.append(filter_aggregate)
+            # Store aggregate by their ids in order to search through id
+            self._aggregate_ids[get_aggregate_id(filter_aggregate)] = filter_aggregate
+
+
+class _DefaultAggregateStore:
+    """This class holds the information of the project associated with
+    an operation function aggregated by the default aggregates i.e.
+    :py:class:`aggregator.groupsof(1)`.
+
+    When iterated through it's instance, it yields a tuple of a single job from
+    the Project.
+
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, project):
+        self._project = project
+
+    def __iter__(self):
+        for job in self._project:
+            yield (job,)
+
+    def __getitem__(self, id):
+        "Return a tuple of a single job via job id."
+        try:
+            return (self._project.open_job(id=id),)
+        except KeyError:
+            raise LookupError(f"Did not find job with id {id}.")
+
+    def __contains__(self, aggregate):
+        """Return whether the job is present in the project associated with this
+        instance of :py:class:`_DefaultAggregateStore`.
+        """
+        # signac-flow internally assumes every aggregate to be a tuple.
+        # Hence this method will also get a tuple as an input.
+        return len(aggregate) == 1 and aggregate[0] in self._project
+
+    def __len__(self):
+        return len(self._project)
+
+    def __eq__(self, other):
+        return type(self) == type(other) and \
+               self._project == other._project
+
+    def __hash__(self):
+        blob = self._project.get_id()
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _register_aggregates(self, project):
+        """We have to store self._project when this method is invoked
+        This is because we will then iterate over that project in
+        order to return an aggregates of one.
+        """
+        self._project = project
+
+
+def get_aggregate_id(jobs):
+    """Generate hashed id for an aggregate of jobs
+
+    :param jobs:
+        The signac job handles
+    :type jobs:
+        tuple
+    """
+    if len(jobs) == 1:
+        return jobs[0].get_id()  # Return job id as it's already unique
+
+    blob = ','.join((job.get_id() for job in jobs))
+    return f'agg-{sha1(blob.encode()).hexdigest()}'
```

Comment:
Since this pattern of string[->sequence of strings]->blob->hash is very common, can you define a simple internal function `_hash` that we can call everywhere? That will ensure consistency in our hashing and simplify any future changes (for instance, to which hash function we call).